### PR TITLE
gre: refactor to allow using L1 and L2 simultaneously

### PR DESCRIPTION
**Pull request by tmigone**
This PR adds support to GRE to allow using L1 and L2 chains simultaneously. This greatly simplifies interacting with the protocol. 

Note that this is on hardhat's roadmap (see https://github.com/NomicFoundation/hardhat/issues/2797) but it's going to take a while before it's ready. 

Relevant docs with more information:
- https://www.notion.so/edgeandnode/Multilayer-support-for-GRE-b2e41614822f453d89f47fa83afe9b95
- README.md file in `gre` folder

### CODE REVIEW

**Comment by abarmat** [comment]
Looking good

File: gre/config.ts [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+  const l2ChainId = isHHL2 ? mainChainId : secondaryChainId
+
+  logDebug(`L1 chain id: ${l1ChainId} - Is HHL1: ${isHHL1}`)
+  logDebug(`L2 chain id: ${l2ChainId} - Is HHL2: ${isHHL2}`)
+
+  return {
+    l1ChainId,
+    l2ChainId,
+    isHHL1,
+    isHHL2,
```
Comment (abarmat) [inline gre/config.ts:87]: Is the isHHL2 ever used? as the first isHHL1 is binary. Or this is just included for symmetry?

File: package.json [hunk 2]
```diff
@@ -91,6 +91,7 @@
     "test:e2e": "scripts/e2e",
     "test:gas": "RUN_EVM=true REPORT_GAS=true scripts/test",
     "test:coverage": "scripts/coverage",
+    "test:gre": "cd gre && mocha --exit --recursive 'test/**/*.test.ts' && cd ..",
```
Comment (abarmat) [inline package.json:94]: 💯

**Review by abarmat — APPROVED** [review]
Congrats @tmigone this is looking great

**Comment by abarmat** [comment]
@tmigone what are the empty config files under `gre/test/files`, are those for existence tests?

File: .mocharc.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "require": "ts-node/register/files",
```
[the remark this answers is not in the document]
Comment (tmigone) [reply .mocharc.json:2]: I knew you would ask ;)

The plugin tests run with mocha (and not with hh test), this is needed for mocha to support typescript files. 

On a more general note, I based the GRE folder on hardhat's plugin boilerplate https://github.com/NomicFoundation/hardhat-ts-plugin-boilerplate/, once GRE is mature I think it makes sense to fork it into it's own repository so each repo is nice and tidy and not polluted with too much stuff.

File: gre/config.ts (hunk 1 above)
Comment (tmigone) [reply gre/config.ts:87]: No, `isHHL2` is not used. Just included as you say for symmetry. Perhaps we can remove it and rename the `isHHL1` variable to something else but I struggled to find a good name.

**Comment by tmigone** [comment]
> @tmigone what are the empty config files under `gre/test/files`, are those for existence tests?

Exactly. The tests are mainly to verify the plugin does path resolution correctly since there are a lot of cases to evaluate (and files need to exist to avoid erroring out). The contents of the files don't really matter as here we don't go that far so I figured it was better to leave them empty.

**Comment by abarmat** [comment]
Fixes GRT-8

**Pull request merged**